storage/local/storage.go

```diff
@@ -716,20 +697,20 @@ func (s *memorySeriesStorage) preloadChunksForRange(
 	if !ok {
 		has, first, last, err := s.persistence.hasArchivedMetric(fp)
 		if err != nil {
-			return nil, err
+			return nil, nopIter, err
 		}
 		if !has {
 			s.invalidPreloadRequestsCount.Inc()
-			return nil, nil
+			return nil, nopIter, nil
 		}
-		if from.Add(-stalenessDelta).Before(last) && through.Add(stalenessDelta).After(first) {
```

Comment:
That is because we switched, some time ago, to a behavior where we only ever return samples _before_ the queried timestamp. We didn't change it in the storage, but then actively threw away the future sample later. In this PR, that useless work is removed by replacing `ValueAtTime(model.Time) []model.SamplePair` with `ValueAtOrBeforeTime(model.Time) model.SamplePair` and removing the throwing-away code in `chooseClosestBefore()`.